domain/modelagent/service/service_test.go

```diff
@@ -0,0 +1,33 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+
+	"github.com/juju/testing"
+	"go.uber.org/mock/gomock"
+	gc "gopkg.in/check.v1"
+)
+
+type suite struct {
```

Comment:
Can you please add a model not found test and also add a test asserting the happy path.

Reply:
I can, but the problem is that all of that logic is in the state layer, so it would be mocked out in this test, so effectively the test is not testing anything.

Modulo setup, it would look something like

```go
s.state.EXPECT().GetModelAgentVersion(gomock.Any(), modelID).Return(expectedVer, nil)
ver, err := svc.GetModelAgentVersion(context.Background(), "invalid")
c.Check(err, jc.ErrorIsNil)
c.Check(ver, jc.DeepEquals, expectedVer)
```

so basically all it tests is that the mock values provided to the function are the same ones returned by the function. Hence I don't really see the value in adding a test here. But I'm happy to do it if you insist.

Reply:
Completely agree and hence my problem with mocks and why I really don't like them.

The thing to remember when testing is that you are currently testing the interface layer of the caller to the function and the contract that we have established. We want to assert that the contract is being upheld so that if the internal implementation of the function changes over time the person making the change can have some assurance that they haven't broken fundamental logic in Juju and possibly change the way the system operates.

Because mocks force us to write tests based on the internal implementation rather then the contracts we have no option but to pass the error via the state mocks. But that is ok because we still want to see that this error bubbles all the way up to satisfy the contract that we are testing at the end of the day.

The test is not about asserting what we can all read as correct logic today but about making sure the contract is maintained over time and not broken with change to the api.

Hope my mad testing rants make sense.

Reply:
Another way to think about this. If you were doing test driven development you would be writing the tests from a black box point of view asserting the contract. Because we use mocks in this way it forces us to write tests based on internal implementation.

I have had a few cracks at trying to get Juju away from this practice but for today we are here and still need to assert correctness over time.

Reply:
fair point, I've added these tests now